lib/geordi/settings.rb

```diff
@@ -78,36 +78,22 @@ def read_settings
       end
 
       if File.exists?(firefox_path)
-        Geordi::Interaction.warn "Unsupported setting \".firefox-version\". Please remove the file in #{firefox_path}."
-        exit 1
+        Interaction.warn "Unsupported config file \".firefox-version\". Please remove #{firefox_path}."
       end
 
       @global_settings = global_settings || {}
       @local_settings = local_settings || {}
     end
 
     def check_for_invalid_keys(settings = [], allowed_keys, file)
-      unless settings.empty?
-        invalid_keys = settings.keys - allowed_keys
-        unless invalid_keys.empty?
-          Geordi::Interaction.warn "Geordi detected unknown keys in #{file}.\n"
-
-          invalid_keys.sort.each do |key|
-            if key == :use_vnc
-              puts "* use_vnc: Unsupported setting \"use_vnc\". Please remove this setting from #{file}"
-            else
-              puts "* #{key}"
-            end
-          end
-
-          puts "\nAllowed keys are:"
-          allowed_keys.sort.each do |key|
-            puts "* #{key}"
-          end
-          puts
-
-          exit 1
-        end
+      return if ENV['GEORDI_INVALID_SETTINGS_WARNED'].to_s.include? file
+      return if settings.empty?
+
+      invalid_keys = settings.keys - allowed_keys
+      unless invalid_keys.empty?
+        Interaction.warn "Unknown settings in #{file}: #{invalid_keys.join(", ")}"
+        Interaction.note "Supported settings in #{file} are: #{allowed_keys.join(", ")}"
+        ENV['GEORDI_INVALID_SETTINGS_WARNED'] = "#{ENV['GEORDI_INVALID_SETTINGS_WARNED']} #{file}"
```

Comment:
Right, we did not consider that there are multiple config files. Your fix will probably work, but it feels hacky. Also, you did not add a test (should have been a spec). I believe, it shows that the "did we already warn" check is placed at the wrong location. So please

* Keep the simplification of this method (`return if settings.empty?`)
* Extract L70-82 to `validated_configuration`
* In this new method, do the check with a boolean `ENV['GEORDI_INVALID_SETTINGS_WARNED']`

Reply:
I don't see the benefit of extracting L70-82 into `validated_configuration`. Wouldn't it be possible to check for the boolean in the read_settings function? That way the paths above (global/local/firefox_path) are in the correct scope:

```
def read_settings
    global_path = GLOBAL_SETTINGS_FILE_NAME
    local_path = LOCAL_SETTINGS_FILE_NAME
    firefox_path = File.expand_path("../../.firefox-version")

    unless ENV['GEORDI_INVALID_SETTINGS_WARNED']
        if File.exists?(global_path)
            global_settings = YAML.safe_load(File.read(global_path)) || {}
            check_for_invalid_keys(global_settings, ALLOWED_GLOBAL_SETTINGS, global_path)
        end

        if File.exists?(local_path)
            local_settings = YAML.safe_load(File.read(local_path)) || {}
            check_for_invalid_keys(local_settings, ALLOWED_LOCAL_SETTINGS, local_path)
        end
    end

    Interaction.warn "Unsupported config file \".firefox-version\". Please remove #{firefox_path}." if File.exists?(firefox_path)

    @global_settings = global_settings || {}
    @local_settings = local_settings || {}
end
```

After that, we just set ENV['GEORDI_INVALID_SETTINGS_WARNED']  to 'true' whenever we warn the user about the settings.

Reply:
Right, this is similar to what I asked you. However, the solution I suggested has three advantages:

* Extracting the settings validation into its own method improves code structure and readability
* Including the "firefox version" check in that function ensures the warning will also printed once at most, which proves this is the right location to set and read the ENV variable
* It narrows the scope of that ENV variable, because you can simply `return` at the beginning of that method if it is set, and set it at the end of the method. Other developers have less to wonder about that variable.

Please also extend the Cucumber test for this feature so it ensures a consecutive execution of Geordi (i.e. the cucumber command) will warn again, i.e. consecutive warnings are only suppressed for the current execution, and this is not persisted.